packages/dashboard-message-bus/lib/DashboardMessageBus.ts

```diff
@@ -0,0 +1,194 @@
+import WebSocket from "isomorphic-ws";
+import delay from "delay";
+import { EventEmitter } from "events";
+import {
+  base64ToJson,
+  broadcastAndAwaitFirst,
+  broadcastAndDisregard,
+  createMessage,
+  jsonToBase64,
+  startWebSocketServer
+} from "./utils";
+import { promisify } from "util";
+
+interface UnfulfilledRequest {
+  publisher: WebSocket;
+  data: WebSocket.Data;
+}
+
+export class DashboardMessageBus extends EventEmitter {
+  private publishServer: WebSocket.Server;
+  private subscribeServer: WebSocket.Server;
+  private publishers: WebSocket[] = [];
+  private subscribers: WebSocket[] = [];
+
+  private unfulfilledRequests: Map<string, UnfulfilledRequest> = new Map([]);
+
+  constructor(
+    public publishPort: number,
+    public subscribePort: number,
+    public host: string = "localhost"
+  ) {
+    super();
+  }
+
+  async start() {
+    this.subscribeServer = await startWebSocketServer({
+      host: this.host,
+      port: this.subscribePort
+    });
+
+    this.subscribeServer.on("connection", (newSubscriber: WebSocket) => {
+      newSubscriber.on("close", () => {
+        this.logToPublishers("Subscriber disconnected", "connections");
+
+        this.subscribers = this.subscribers.filter(
+          subscriber => subscriber !== newSubscriber
+        );
+
+        this.terminateIfNoConnections();
+      });
+
+      // Require the subscriber to send a message *first* before being marked as ready
+      newSubscriber.once("message", () => this.addNewSubscriber(newSubscriber));
+    });
+
+    this.publishServer = await startWebSocketServer({
+      host: this.host,
+      port: this.publishPort
+    });
+
+    this.publishServer.on("connection", (newPublisher: WebSocket) => {
+      newPublisher.on("message", (data: WebSocket.Data) => {
+        this.processRequest(newPublisher, data, this.subscribers);
+      });
+
+      newPublisher.on("close", () => {
+        this.logToPublishers("Publisher disconnected", "connections");
+
+        this.publishers = this.publishers.filter(
+          publisher => publisher !== newPublisher
+        );
+
+        this.clearRequestsForPublisher(newPublisher);
+        this.terminateIfNoConnections();
+      });
+
+      this.logToPublishers("Publisher connected", "connections");
+
+      this.publishers.push(newPublisher);
+    });
+  }
+
+  async ready() {
+    if (this.subscribers.length > 0) return;
+    await delay(1000);
+    await this.ready();
+  }
```

Comment:
`while` blocks the event loop. `clearInterval` can offer a way to poll without blocking or affecting the callstack in this way.

If it has already been explained wave me off - I'm late to the review. But what is the need to do the `delay(1000)`? I would expect it to imply there is a race condition?

Reply:
So the message bus is "ready" when it has at least 1 subsciber to send things to. Essentially this function polls every second to see if it's "ready" now.

Reply:
As for the event loop blocking. I *think* that blocking the event loop is what we want to do here. This `ready()` promise is used to block the sending of requests until it can send them somewhere.

Reply:
I wonder if this timeout is something that should be configurable. What's the downside of blocking here? I can't really think of one, other than the initial one second might be slow? But I doubt it.

Anyway, "able to actually send a request somewhere" is a decent definition of "ready" if you ask me :)